### ISSUE TITLE
[clang] __STDC_NO_THREADS__ is no longer necessary for VS 2022 17.8

### ISSUE THREAD

**Issue by latal-1** [llvm/llvm-project#115529]
Starting with [VS 2022 17.8](https://learn.microsoft.com/en-us/visualstudio/releases/2022/release-notes-v17.8), the `__STDC_NO_THREADS__` is no longer relevant due to the addition of thread library

Сommit with the addition of a macro https://github.com/llvm/llvm-project/commit/bbc690c57213054907284d8964dc0487d38fc57a

**Comment by AaronBallman** [comment]
17.7: defines `__STDC_NO_THREADS__` does not supply `<threads.h>`
17.8: does not define `__STDC_NO_THREADS__` does not supply `<threads.h>`
17.9: does not define `__STDC_NO_THREADS__`, does supply `<threads.h>`

https://godbolt.org/z/f8san8Ebe

Part of the challenge is that the CRT is defined at runtime rather than at compile time, so it's hard to know whether the 17.8 behavior is because of an old CRT or not.

**Comment by AaronBallman** [comment]
This seems like a reasonably good first issue.

In `addVisualCDefines()` in `clang/lib/Basic/Targets/OSTargets.cpp`, guard the call to `Builder.defineMacro("__STDC_NO_THREADS__");` on the version of Visual Studio we're aiming to be compatible with. You'll likely need to add a new enumeration to `MSVCMajorVersion` to represent MSVC 17.8 (`1938` would be the value for the enumeration).

**Comment by dipeshs809** [comment]
Hello, @AaronBallman, I'm new to contributing to LLVM and would like to pick this up as my first issue to work on. Could you please assign it to me.

**Comment by AaronBallman** [comment]
> The current enumeration are ... MSVC2017_7 = 1914 and MSVC2019 = 1920... , hence as per my understanding, enum for MSVC2017_8 should be kept as MSVC2017_8 = 1916 instead of MSVC2017_8 = 1938.

From https://learn.microsoft.com/en-us/cpp/overview/compiler-versions?view=msvc-170

Visual Studio 2017 (15.7) has `_MSC_VER` as `1914`
Visual Studio 2022 (17.8) has `_MSC_VER` as `1938`

So for our enumeration, `MSVC2017_7` is correct and we need to add `MSVC2022_8` whose value is `1938`.

> Also as per the CRT https://godbolt.org/z/f8san8Ebe, guard should be like this.
> Define macro if MSCompatibilityVersion <= MSVC2017_8 

I'm not entirely certain of what's going on there. CC @CaseyCarter for extra help on this. Basically, does VS 17.8 support `threads.h` (so godbolt is in a confused state in terms of the CRT headers) or does 17.8 not have `threads.h` but fails to define `__STDC_NO_THREADS__` (so it's an MSVC bug and we shouldn't aim to be compatible with it)?

**Comment by CaseyCarter** [comment]
> I'm not entirely certain of what's going on there. CC @CaseyCarter for extra help on this. Basically, does VS 17.8 support `threads.h` (so godbolt is in a confused state in terms of the CRT headers) or does 17.8 not have `threads.h` but fails to define `__STDC_NO_THREADS__` (so it's an MSVC bug and we shouldn't aim to be compatible with it)?

IIRC 17.8 is broken. We went to all the trouble to implement the thread support library, and we got the compiler to stop defining `__STDC_NO_THREADS__`, but we didn't tick the right boxes to actually ship `<threads.h>`. Since it wasn't a regression, we couldn't convince the Powers that Be to ship the new header in a servicing release. VS2022 17.9 is the first release with usable C11 threads. (Unless you consider pasting the contents of the header into your program to be "usable": https://www.godbolt.org/z/ev6cvhMef).

@barcharcraz may be able to add some details.

**Comment by AaronBallman** [comment]
Thank you @CaseyCarter, that's super helpful!

@dipeshs809 -- based on this, you should create `MSVC2022_9` whose value is `1939` and test against that (please be sure the test coverage for the changes tests `-fms-compatibility-version=` with the values `19.39` (won't have macro defined) and `19.38` (will have macro defined).

**Comment by barcharcraz** [comment]
> > I'm not entirely certain of what's going on there. CC @CaseyCarter for extra help on this. Basically, does VS 17.8 support `threads.h` (so godbolt is in a confused state in terms of the CRT headers) or does 17.8 not have `threads.h` but fails to define `__STDC_NO_THREADS__` (so it's an MSVC bug and we shouldn't aim to be compatible with it)?
> 
> IIRC 17.8 is broken. We went to all the trouble to implement the thread support library, and we got the compiler to stop defining `__STDC_NO_THREADS__`, but we didn't tick the right boxes to actually ship `<threads.h>`. Since it wasn't a regression, we couldn't convince the Powers that Be to ship the new header in a servicing release. VS2022 17.9 is the first release with usable C11 threads. (Unless you consider pasting the contents of the header into your program to be "usable": https://www.godbolt.org/z/ev6cvhMef).
> 
> @barcharcraz may be able to add some details.

That's about right. 

something similar to `#if defined(__has_include) && __has_include(<threads.h>) && !defined(__STDC_NO_THREADS__) && __STDC_VERSION__ >= 201112L` may work, but `__has_include` isn't required to be supported prior to C23

**Comment by barcharcraz** [comment]
and note that just checking for `!defined(__STDC_NO_THREADS__)` isn't enough anyway because that macro was added in C11! Similarly just checking for `<threads.h>` isn't sufficient because most implementations don't really have a way to "hide" headers based on standard mode and `threads.h` might not compile in C89 or C99 mode (ours does not)

**Comment by dipeshs809** [comment]
Thanks @barcharcraz, @CaseyCarter and @AaronBallman. Having all the content digested, we should define the macro `__STDC_NO_THREADS__` in OSTargets.cpp only when:

- The MSVC version is less than 17.9.
- The C standard is C11 or newer `(__STDC_VERSION__ >= 201112L)`.
- `<threads.h>` is unavailable `(!__has_include(<threads.h>))`.

**Comment by CaseyCarter** [comment]
> Having all the content digested, we should define the macro `__STDC_NO_THREADS__` in OSTargets.cpp only when:
> 
> * The MSVC version is less than 17.9.
> * The C standard is C11 or newer `(__STDC_VERSION__ >= 201112L)`.
> * `<threads.h>` is unavailable `(!__has_include(<threads.h>))`.

Since `__STDC_NO_THREADS__` is a reserved identifier, we may define it in earlier language modes despite that we're not required to do so. We _must_ define it when `<threads.h>` is unavailable. We probably _should_ define it when the MSVC version is less than 17.9; lack of `<threads.h>` should catch this case anyway, but it doesn't hurt to be a little paranoid. We _may_ define it when `__STDC_VERSION__ < 201112L` to help users who aren't expert in how to detect thread support.

**Comment by dipeshs809** [comment]
Hi, @AaronBallman 

I've raised a PR to close the issue. All test cases ran, and the build was also generated successfully. Could you please review it once.

**Comment by erichkeane** [comment]
@dipeshs809 : I see this patch is waiting on you to revise your patch, are you going to continue to work on this, or should we mark this as unassigned?

**Comment by dipeshs809** [comment]
> [@dipeshs809](https://github.com/dipeshs809) : I see this patch is waiting on you to revise your patch, are you going to continue to work on this, or should we mark this as unassigned?

@erichkeane My apology for the delay, I'll be fixing this patch in a few days.
